### PR TITLE
diff: make diff a package && can ignore some columns

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
1. make diff a package
2. can ignore some columns, will not check these columns' data
3. add remove columns config, will not check these columns' data, and will not use these columns as split field or order by key.
4. can ignore check table's struct or table's data

### CODE REVIEW

File: sync_diff_inspector/util.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+)
+
+func schemaStr(instanceID, schema string) string {
```
Comment (IANTHEREAL) [inline sync_diff_inspector/util.go]: is `schemaName` better?

File: pkg/diff/util.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+			return false, nil
+		}
+	}
+
+	// check index
+	if len(tableInfo1.Indices) != len(tableInfo2.Indices) {
+		return false, nil
+	}
+
+	for i, index := range tableInfo1.Indices {
```
Comment (IANTHEREAL) [inline pkg/diff/util.go]: is this check too strict？ do we need compatible with indexes in different orders？

File: pkg/diff/util.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	}
+	for i := 0; i < len(str1); i++ {
+		if str1[i] != str2[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func equalTableInfo(tableInfo1, tableInfo2 *model.TableInfo) (bool, error) {
```
Comment (IANTHEREAL) [inline pkg/diff/util.go]: it's better to put it into dbutil pkg, check also does same thing.

File: pkg/diff/util.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+			if col.Name.O != index2.Columns[j].Name.O {
+				return false, nil
+			}
+		}
+	}
+
+	return true, nil
+}
+
+func ignoreColumns(tableInfo *model.TableInfo, columns []string) *model.TableInfo {
```
Comment (IANTHEREAL) [inline pkg/diff/util.go]: I don't think we need to change `tableinfo`,  e.g. a table with a unique key (id, name,age), if we remove id, name and age maybe not unique, how to sort them?

File: pkg/diff/util.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	if len(columns) == 0 {
+		return tableInfo
+	}
+	ignoreColMap := sliceToMap(columns)
+	for i, index := range tableInfo.Indices {
+		for j, col := range index.Columns {
+			if _, ok := ignoreColMap[col.Name.O]; ok {
+				index.Columns = append(index.Columns[:j], index.Columns[j+1:]...)
+				if len(index.Columns) == 0 {
+					tableInfo.Indices = append(tableInfo.Indices[:i], tableInfo.Indices[i+1:]...)
```
Comment (IANTHEREAL) [inline pkg/diff/util.go:51]: put it out of for loop at L81

File: sync_diff_inspector/util.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	return fmt.Sprintf("%s|%s", instanceID, schema)
+}
+
+func sliceToMap(slice []string) map[string]interface{} {
```
Comment (IANTHEREAL) [inline sync_diff_inspector/util.go]: `pkg/diff/util` also has same function

File: pkg/diff/chunk.go [hunk 7]
```diff
@@ -62,14 +62,14 @@ func newChunkRange(begin, end interface{}, containBegin, containEnd, noBegin, no
 		noEnd:        noEnd,
 	}
 }
-func getChunksForTable(db DBConfig, table *TableConfig, column *model.ColumnInfo, chunkSize, sample int) ([]chunkRange, error) {
+func getChunksForTable(table *TableInstance, column *model.ColumnInfo, chunkSize, sample int, limits string) ([]chunkRange, error) {
```
Comment (IANTHEREAL) [inline pkg/diff/chunk.go]: unify name, s/limits/range

File: pkg/diff/diff.go [hunk 8]
```diff
@@ -0,0 +1,532 @@
...
+	return structEqual, dataEqual, nil
+}
+
+// CheckTableStruct checks table's struct
+func (t *TableDiff) CheckTableStruct(ctx context.Context) (bool, error) {
+	tableInfo, err := dbutil.GetTableInfoWithRowID(ctx, t.TargetTable.Conn, t.TargetTable.Schema, t.TargetTable.Table, t.UseRowID)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	log.Infof("")
```
Comment (IANTHEREAL) [inline pkg/diff/diff.go]: any useful?

**Comment by IANTHEREAL** [comment]
@zhouqiang-cl @csuzhangxc PTAL

File: pkg/diff/chunk.go (hunk 7 above)
Comment (WangXiangUSTC) [reply pkg/diff/chunk.go]: range is a keyword in golang, so I use limits

File: pkg/diff/util.go (hunk 4 above)
Comment (WangXiangUSTC) [reply pkg/diff/util.go]: I add a remove-columns config, if user set ignore-columns, will not check these columns' data, if set remove-columns, will remove these columns from table info.

File: pkg/diff/util.go (hunk 3 above)
Comment (WangXiangUSTC) [reply pkg/diff/util.go]: done

File: pkg/diff/util.go (hunk 2 above)
Comment (WangXiangUSTC) [reply pkg/diff/util.go]: done

File: pkg/diff/util.go (hunk 5 above)
Comment (WangXiangUSTC) [reply pkg/diff/util.go:51]: only after remove columns, may index had 0 columns

File: pkg/dbutil/table.go [hunk 9]
```diff
@@ -280,3 +280,48 @@ func setImplicitColumn(table *model.TableInfo) {
 	}
 	table.Indices = []*model.IndexInfo{newIndex}
 }
+
+// EqualTableInfo returns true if this two table info have same columns and indices
+func EqualTableInfo(tableInfo1, tableInfo2 *model.TableInfo) (bool, error) {
```
Comment (zhouqiang-cl) [inline pkg/dbutil/table.go]: when return error

File: pkg/diff/util.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+	for j := 0; j < num; j++ {
+		r := j + rand.Intn(total-j)
+		totalArray[j], totalArray[r] = totalArray[r], totalArray[j]
+	}
+
+	return totalArray[:num]
+}
+
+func needQuotes(ft types.FieldType) bool {
```
Comment (zhouqiang-cl) [inline pkg/diff/util.go:88]: number & float can use quota too

File: sync_diff_inspector/util.go [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+)
+
+func schemaName(instanceID, schema string) string {
```
Comment (zhouqiang-cl) [inline sync_diff_inspector/util.go:20]: where use this function

**Review by zhouqiang-cl — COMMENTED** [review]
rest LGTM

File: pkg/dbutil/table.go (hunk 9 above)
Comment (WangXiangUSTC) [reply pkg/dbutil/table.go]: I will remove error

File: pkg/diff/util.go (hunk 10 above)
Comment (WangXiangUSTC) [reply pkg/diff/util.go:88]: I will remove this function

File: sync_diff_inspector/util.go (hunk 11 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/util.go:20]: it is used for generate a label for schema,  you can search the code

File: pkg/diff/chunk.go (hunk 7 above)
Comment (csuzhangxc) [reply pkg/diff/chunk.go]: so, is it possible to put `limits` into `table`

File: pkg/diff/util.go (hunk 10 above)
Comment (WangXiangUSTC) [reply pkg/diff/util.go:88]: but "2"> "11" will return true, I will keep this function

File: pkg/diff/chunk.go (hunk 7 above)
Comment (WangXiangUSTC) [reply pkg/diff/chunk.go]: table is a  instance for table, seems not suitable

File: pkg/diff/util.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+import (
+	"math/rand"
+
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb-tools/pkg/dbutil"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/types"
+)
+
+func equalStrings(str1, str2 []string) bool {
```
Comment (csuzhangxc) [inline pkg/diff/util.go:25]: where use this function? and why not use `==` simply

Comment (WangXiangUSTC) [reply pkg/diff/util.go:25]: slice can only be compared to nil. this function is used to check whether two schema have same tables, I will use this function later

Comment (csuzhangxc) [reply pkg/diff/util.go:25]: wow, my mistake

File: pkg/diff/util.go [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+func removeColumns(tableInfo *model.TableInfo, columns []string) *model.TableInfo {
+	if len(columns) == 0 {
+		return tableInfo
+	}
+	removeColMap := SliceToMap(columns)
+	for i, index := range tableInfo.Indices {
+		for j, col := range index.Columns {
+			if _, ok := removeColMap[col.Name.O]; ok {
+				index.Columns = append(index.Columns[:j], index.Columns[j+1:]...)
```
Comment (csuzhangxc) [inline pkg/diff/util.go:48]: is is possible `slice bounds out of range` for `index.Columns[j+1:]` ?

**Comment by csuzhangxc** [comment]
rest LGTM

**Review by zhouqiang-cl — APPROVED** [review]
LGTM

File: pkg/diff/util.go (hunk 13 above)
Comment (WangXiangUSTC) [reply pkg/diff/util.go:48]: yes, this is a bug, I fix it

File: pkg/diff/util.go [hunk 14]
```diff
@@ -38,21 +38,28 @@ func removeColumns(tableInfo *model.TableInfo, columns []string) *model.TableInf
 	if len(columns) == 0 {
 		return tableInfo
 	}
+
 	removeColMap := SliceToMap(columns)
-	for i, index := range tableInfo.Indices {
-		for j, col := range index.Columns {
+	for i := 0; i < len(tableInfo.Indices); i++ {
```
Comment (IANTHEREAL) [inline pkg/diff/util.go:43]: I prefer to construct a new columns. but we can keep it

**Comment by IANTHEREAL** [comment]
LGTM

**Pull request merged**